### PR TITLE
JavaFX Incubator Modules

### PR DESCRIPTION
**Pull request by kevinrushforth**
This PR contains a proposal for [JavaFX Incubator Modules](https://github.com/kevinrushforth/jfx/blob/javafx.incubator/INCUBATOR-MODULES.md), along with an example of how such a module might be implemented.

NOTE: This PR will remain in Draft. It is intended only as supporting material for the proposal.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1375/head:pull/1375` \
`$ git checkout pull/1375`

Update a local copy of the PR: \
`$ git checkout pull/1375` \
`$ git pull https://git.openjdk.org/jfx.git pull/1375/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1375`

View PR using the GUI difftool: \
`$ git pr show -t 1375`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1375.diff">https://git.openjdk.org/jfx/pull/1375.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
I pushed the following updates:

- Added a utility to print an incubator warning
- Updated the Incubator Modules JEP:
    - Change `javafx.incubator` to `jfx.incubator`
    - Say that a JEP is needed to finalize or drop an incubating feature
    - Clarify that a feature should not incubate indefinitely
- Updated the instructions to create a new incubator module. As part of this, I marked the changes in the JavaFX core (e.g., in `build.gradle`) with one of the following two comments:
    - `//TODO: incubator dependency` : Changes that will be needed in the core in support of incubator modules in general, which include the new utlity class and many of the changes in `build.gradle`; these changes will need go into mainline (via a separate enhancement) ahead of any specific incubator module
    -  `//TODO: incubator template` : Identifies modifications needed in the core for a specific incubator module; these changes would be part of a PR for that specific incubator

File: modules/javafx.base/src/main/java/com/sun/javafx/ModuleUtil.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ */
+public class ModuleUtil {
+
+    private static final Set<Module> warnedModules = new HashSet<>();
+    private static final Set<Package> warnedPackages = new HashSet<>();
+
+    private static final Module MODULE_JAVA_BASE = Module.class.getModule();
+
+    @SuppressWarnings("removal")
+    public static void incubatorWarning() {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.base/src/main/java/com/sun/javafx/ModuleUtil.java:45]: perhaps we can add a javadoc to this method explaining why it is here and giving a usage example?

**Comment by andy-goryachev-oracle** [comment]
Incubator warning works as expected:

```
WARNING: Using incubator modules: jfx.incubator.richtext
```

File: modules/javafx.base/src/main/java/com/sun/javafx/ModuleUtil.java (hunk 1 above)
Comment (kevinrushforth) [reply modules/javafx.base/src/main/java/com/sun/javafx/ModuleUtil.java:45]: Good idea.

**Comment by kevinrushforth** [comment]
> BTW, should we have a separate, real PR with changes _only_ for enabling the functionality (.md file and build.gradle, the latter with sample module changes commented out - maybe with an easy to grep marker for anyone who wants to add a new incubator module)?

Yes, something along those lines is in the works.

I've already identified many of the changes that would be part of this separate Enhancement, and marked them with `//TODO: incubator dependency` -- see the [above comment](#issuecomment-1974026095).

I'm not sure yet what the best way is to show where to add the logic for a new incubator module. At least having an easy-to-find "incubator modules go here" comment would be helpful and seems like a good idea. As you suggest, I could go farther than that and add commented-out build logic for the sample module...need to think about that one. I would not be in favor of including the sample module, though, so I'm leaning towards just adding the appropriate markers in build.gradle, but not any actual build logic for an incubator modules. Anyone doing an incubator module would need to grab the sample patch anyway.

File: build.gradle [hunk 2]
```diff
@@ -3863,7 +3976,18 @@ project(":systemTests") {
...
     }
 
-    def dependentProjects = [ 'base', 'graphics', 'controls', 'media', 'web', 'swing', 'fxml' ]
+    // TODO: incubator dependency
+    def dependentProjects = [
+        'base',
+        'graphics',
+        'controls',
+        // TODO: incubator template
+        'incubator.myfeature',
```
Comment (andy-goryachev-oracle) [inline build.gradle:3985]: this change fixed the jenkins build of the rich text area incubator, thank you!

**Comment by kevinrushforth** [comment]
> @kevinrushforth kevinrushforth force-pushed the javafx.incubator branch from 1f567a9 to 4bb1b59 
now

I pushed the wrong merge commit, so corrected it before (I hope) anyone other than Skara noticed. I wouldn't do this except that this is a Draft PR that I don't intend to ever make `rfr`.

I plan to close this Draft PR soon anyway, and create a new PR from a clean branch.

**Comment by kevinrushforth** [comment]
Keep alive

**Comment by kevinrushforth** [comment]
I am closing this PR in favor of #1617 (Draft PR for JavaFX Incubator Modules) and #1616 (PR to add support for JavaFX Incubator Modules)

**Pull request closed without merging**